File: backend_v2/app/core/document_service.py
from app.utils.pdf_utils import extract_text_from_pdf
from app.core.vectorstore import VectorStore
from datetime import datetime
import tiktoken
from openai import OpenAI

vector_store = VectorStore()
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

def chunk_text(text: str, max_tokens: int = 500):
    # Simples chunk por tokens aproximados
    chunks = []
    current = []
    tokens_count = 0
    encoder = tiktoken.get_encoding("cl100k_base")
    for line in text.split("\n"):
        line_tokens = len(encoder.encode(line))
        if tokens_count + line_tokens > max_tokens:
            chunks.append("\n".join(current))
            current = []
            tokens_count = 0
        current.append(line)
        tokens_count += line_tokens
    if current:
        chunks.append("\n".join(current))
    return chunks

def ingest_pdf(file_bytes: bytes, filename: str):
    text = extract_text_from_pdf(file_bytes)
    chunks = chunk_text(text)
    for i, chunk in enumerate(chunks, start=1):
        embedding = client.embeddings.create(
            model="text-embedding-3-small",
            input=chunk
        ).data[0].embedding
        metadata = {
            "file_name": filename,
            "chunk_index": i,
            "timestamp": datetime.utcnow().isoformat()
        }
        vector_store.add(chunk, metadata, embedding)
    return {"message": f"Documento '{filename}' processado com sucesso.", "chunks_count": len(chunks)}

File: backend/chat/src/services/DocumentService.ts
import { extractTextFromPdf } from '../utils/pdfUtils.js';
import VectorStore from '../core/VectorStore.js';
import DocumentRepository from '../repositories/DocumentRepository.js';
import { getEmbeddingProvider } from '../core/providers/providerFactory.js';
import {sleep} from "../utils/RequestUtils";
import {logger} from "../utils/logger";

const embeddingProvider = getEmbeddingProvider();
const documentRepository = new DocumentRepository();

// Token counting approximation for JavaScript
function countTokens(text: any) {
  // Approximation: ~4 characters per token (similar to tiktoken)
  return Math.ceil(text.length / 4);
}

function chunkText(text: any, maxTokens: number = 1500): string[] {
  const chunks: string[] = [];
  const lines = text.split('\n');
  let current: string[] = [];
  let tokensCount = 0;

  for (const line of lines) {
    const lineTokens = countTokens(line);
    if (tokensCount + lineTokens > maxTokens) {
      chunks.push(current.join('\n'));
      current = [];
      tokensCount = 0;
    }
    current.push(line);
    tokensCount += lineTokens;
  }

  if (current.length > 0) {
    chunks.push(current.join('\n'));
  }

  return chunks;
}

async function ingestPdf(fileBytes: any, filename: any, userId: number, chatId: any = null) {
  try {
    const pagesText = await extractTextFromPdf(fileBytes);

    const chunks = chunkText(pagesText);

    const documentId = await documentRepository.createDocument(userId, filename);

    const embeddings = await embeddingProvider.embedChunks(chunks);

    const vectorStore = new VectorStore();

    const savePromises = chunks.map((chunk, i) => {
      const metadata = {
        file_name: filename,
        chunk_index: i + 1,
        timestamp: new Date().toISOString(),
        user_id: userId,
        chat_id: chatId,
        document_id: documentId,
      };
      return vectorStore.add(chunk, metadata, embeddings[i]);
    });

    await Promise.all(savePromises);

    return {
      message: `Documento '${filename}' processado com sucesso.`,
      chunks_count: chunks.length,
      user_id: userId,
      document_id: documentId,
    };
  } catch (error: any) {
    logger.error(`Erro ao ingerir PDF: ${error.message}`);
    throw new Error(`Erro ao ingerir PDF: ${error.message}`);
  }
}

export { ingestPdf };

